### PR TITLE
Reenabled param -gen, run miner when flag is on

### PR DESCRIPTION
**Pull request by izolyomi**

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -1464,5 +1470,22 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
     return !fRequestShutdown;
 }
+
+
+UniValue MinerThread(boost::shared_ptr<CReserveScript> coinbaseScript)
+{
+    while (true) {
+        UniValue result = generateBlocks(coinbaseScript, 1, UINT64_MAX, true);
+        // TODO should we do anything with the result except for logging it?
+    }
```
Comment (wigy-opensource-developer) [inline src/init.cpp]: Leaving the mining thread for any reason (like an exception) is a big deal for a deamon started to do mining. I would put everything inside the loop into a try-catch, log the exception, wait a bit and then go back mining.

File: src/init.cpp [hunk 2]
```diff
@@ -1464,5 +1470,22 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     }
 #endif
 
+//    LogPrintf("***** Option '-gen' (mining) is turned %s *****\n",
+//        GetBoolArg("-gen", false) ? "on" : "off" );
+    if (GetBoolArg("-gen", false)) {
```
Comment (wigy-opensource-developer) [inline src/init.cpp]: Maybe we should reassure the user with a log here that we started mining?

File: src/init.cpp [hunk 3]
```diff
@@ -1481,8 +1486,25 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 
 void MinerThread()
 {
-    boost::shared_ptr<CReserveScript> coinbaseScript;
-    GetMainSignals().ScriptForMining(coinbaseScript);
+    // Initialize mining options, especially the mining address
+    string mineToAddressStr = GetArg("-genaddr", "");
```
Comment (wigy-opensource-developer) [inline src/init.cpp]: Parsing and validating the command-line does not seem to belong to this thread. How about turning this method into a method object? Then the parsing could be done in a static factory method and the address could be stored in a field of the object. If there was no error, the thread could be started on an instance method. If there was some error, the factory method could just return null, and the main thread could just go on.

Should you need help in that, just call me on Hangout.

**Comment by izolyomi** [comment]
Mining address parameter is added. It is a mandatory parameter, the user must be aware of his address anyway while we are whitelisting.

File: src/init.cpp (hunk 3 above)
Comment (izolyomi) [inline src/init.cpp]: Bitcoin does parameter handling with a global and static map, it should be thread-safe when you just read. However you're right, it's better outside the thread, will commit changes soon.

**Comment by izolyomi** [comment]
Clarified help message as requested in slack discussion by Daniel.

**Pull request merged**